tests/unit/s2n_psk_test.c

```diff
@@ -194,9 +345,177 @@ int main(int argc, char **argv)
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+
+            EXPECT_OK(s2n_psk_write_binder_list(conn, &client_hello_prefix, &out));
+
+            uint16_t binder_list_size;
+            EXPECT_SUCCESS(s2n_stuffer_read_uint16(&out, &binder_list_size));
+            EXPECT_EQUAL(binder_list_size, s2n_stuffer_data_available(&out));
+
+            uint8_t binder_size;
+            EXPECT_SUCCESS(s2n_stuffer_read_uint8(&out, &binder_size));
+            EXPECT_EQUAL(binder_size, s2n_stuffer_data_available(&out));
+            EXPECT_EQUAL(binder_size, finished_binder.size);
+
+            uint8_t *binder_data;
+            EXPECT_NOT_NULL(binder_data = s2n_stuffer_raw_read(&out, binder_size));
+            EXPECT_BYTEARRAY_EQUAL(binder_data, finished_binder.data, binder_size);
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(s2n_stuffer_free(&out));
+        }
+
+        /* Test s2n_psk_write_binder_list with multiple PSKs */
+        {
+            const uint8_t psk_count = 5;
+
+            struct s2n_connection *conn;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+
+            for (uint8_t i = 0; i < psk_count; i++) {
+                struct s2n_psk *psk;
+                EXPECT_OK(s2n_array_pushback(&conn->psk_params.psk_list, (void**) &psk));
+                EXPECT_SUCCESS(s2n_psk_init(psk, S2N_PSK_TYPE_RESUMPTION));
+                EXPECT_SUCCESS(s2n_psk_new_identity(psk, identity.data, identity.size));
+                EXPECT_SUCCESS(s2n_psk_new_secret(psk, resumption_secret.data, resumption_secret.size));
+            }
+
+            struct s2n_stuffer out;
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+
+            EXPECT_OK(s2n_psk_write_binder_list(conn, &client_hello_prefix, &out));
+
+            uint16_t binder_list_size;
+            EXPECT_SUCCESS(s2n_stuffer_read_uint16(&out, &binder_list_size));
+            EXPECT_EQUAL(binder_list_size, s2n_stuffer_data_available(&out));
+
+            for (uint8_t i = 0; i < psk_count; i++) {
+                uint8_t binder_size;
+                EXPECT_SUCCESS(s2n_stuffer_read_uint8(&out, &binder_size));
+                EXPECT_EQUAL(binder_size, finished_binder.size);
+
+                uint8_t *binder_data;
+                EXPECT_NOT_NULL(binder_data = s2n_stuffer_raw_read(&out, binder_size));
+                EXPECT_BYTEARRAY_EQUAL(binder_data, finished_binder.data, binder_size);
+            }
+
+            EXPECT_EQUAL(s2n_stuffer_data_available(&out), 0);
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(s2n_stuffer_free(&out));
+        }
+
+        /* Test s2n_psk_write_binder_list with multiple hash algs */
+        {
+            struct s2n_connection *conn;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+
+            for (s2n_hash_algorithm hash_alg = S2N_HASH_SHA1; hash_alg < S2N_HASH_SHA384; hash_alg++) {
```

Comment:
Where does it say which hash algorithms can be used for the psk in the rfc?

Reply:
It's just the hash algorithms available from the current TLS1.3 cipher suites and useable with our current TLS1.3 KDF code, which assumes the digest will never exceed the size of sha384's digest. So sha384 SHOULD be allowed, I missed an '=' there, good catch!